### PR TITLE
Create the backend algorithm for brain and grid

### PR DESCRIPTION
**Pull request by RolandNaijuka**
- Re-implement the graph to store grid items' names as vertices, and grid items as labels on the vertices of the graph
- Implement Dijkstra's algorithm to find the shortest distance between the grid item requesting energy, and the one which can give it (requester) energy.
- Implement grid items to have similar functions that return the same types for easy use in the algorithm to decide where to send power or energy
- Implement the batteries to collect and give energy
- Implement the houses to drain the energy
- Implement the brain to create a grid action
- Implement the Biogrid to take action on what the brain returned
- Files that had lots of changes: 
  - Biobrain
    - Includes the algorithm to determine which energy user requires energy and which energy provider can give it
    - [bio-brain](https://github.com/googleinterns/step141-2020/pull/47/files#diff-6c9d5d03ecf99ce1c43cd7e959a16c60)
    - [respective-tests](https://github.com/googleinterns/step141-2020/pull/47/files#diff-30d1ffdd6d67c26d97d822d642e541cb)
  - Biogrid state
    - Includes a graph of the different states of the grid. That is the grid items
    - [bio-grid-state](https://github.com/googleinterns/step141-2020/pull/47/files#diff-0c0222290e605c2f2addd1e6d0bdde35)
    - [repsective-tests](https://github.com/googleinterns/step141-2020/pull/47/files#diff-e3e1f174e4eddc2fa1793fae46dc6712)
  - Biogrid
    - Includes setting and placing the batteries in the grid
    - Includes initializing the `biogrid state`
    - Includes an algorithm which responds to the action returned by the `brain`
    - [biogrid](https://github.com/googleinterns/step141-2020/pull/47/files#diff-24f3ef3a1b95e6c1fc5e3265e6233267)
    - [respective-tests](https://github.com/googleinterns/step141-2020/pull/47/files#diff-75fad9e126df25d1be3b4ba8a1112cb6)

This links to issue #30 and works on parts 4, 5 and last two bullets of 3

### CODE REVIEW

File: biogrid/libs/biogrid-simulator/src/lib/biobrain/biobrain-class.ts
Comment (Lev-Stambler) [inline biogrid/libs/biogrid-simulator/src/lib/biobrain/biobrain-class.ts:210]: Love this! Could you just leave a comment above explaining how this method works

File: biogrid/libs/biogrid-simulator/src/lib/bioenergy-source/bioenergy-source-class.ts
Comment (Lev-Stambler) [inline biogrid/libs/biogrid-simulator/src/lib/bioenergy-source/bioenergy-source-class.ts]: I would look over this again after merging with master, this file name was changed in the last merged PR, now there is a solar panel file as well as a bioenergy source abstract class.

File: biogrid/libs/biogrid-simulator/src/lib/biogrid-state/biogrid-state-class.ts
Comment (Lev-Stambler) [inline biogrid/libs/biogrid-simulator/src/lib/biogrid-state/biogrid-state-class.ts:18]: great change!

File: biogrid/libs/biogrid-simulator/src/lib/biogrid-state/biogrid-state-class.ts
Comment (Lev-Stambler) [inline biogrid/libs/biogrid-simulator/src/lib/biogrid-state/biogrid-state-class.ts:123]: awesome comment, maybe instead of `addEdge`, the name could be `connectNewVertex` or something along those lines. Add edge may imply just adding an edge between 2 vertices

File: biogrid/libs/biogrid-simulator/src/lib/biogrid-state/biogrid-state-class.ts
Comment (Lev-Stambler) [inline biogrid/libs/biogrid-simulator/src/lib/biogrid-state/biogrid-state-class.ts]: nit: format to multiple lines

File: biogrid/libs/biogrid-simulator/src/lib/biogrid-state/biogrid-state-class.ts
Comment (Lev-Stambler) [inline biogrid/libs/biogrid-simulator/src/lib/biogrid-state/biogrid-state-class.ts]: nit: multiple lines

File: biogrid/libs/biogrid-simulator/src/lib/biogrid/biogrid-class.ts
Comment (Lev-Stambler) [inline biogrid/libs/biogrid-simulator/src/lib/biogrid/biogrid-class.ts:41]: again, see new PR merge

File: biogrid/libs/biogrid-simulator/src/lib/biogrid/biogrid-class.ts
Comment (Lev-Stambler) [inline biogrid/libs/biogrid-simulator/src/lib/biogrid/biogrid-class.ts:124]: could this be broken up using helper functions. I.e. if it is an energy user, a function like `takeEnergyUserAction`

File: biogrid/libs/biogrid-simulator/src/lib/biogrid/biogrid-class.ts
Comment (Lev-Stambler) [inline biogrid/libs/biogrid-simulator/src/lib/biogrid/biogrid-class.ts:136]: nit: remove else and continue

File: biogrid/libs/biogrid-simulator/src/lib/config/bio-constants.ts
Comment (Lev-Stambler) [inline biogrid/libs/biogrid-simulator/src/lib/config/bio-constants.ts:18]: See the new PR as well

**Review by Lev-Stambler — CHANGES_REQUESTED** [review]

**Comment by Lev-Stambler** [comment]
Looks great! Just a few things and merging in from the most recent master branch.

File: biogrid/libs/biogrid-simulator/src/lib/biogrid-state/biogrid-state-class.ts
Comment (aosmann7) [inline biogrid/libs/biogrid-simulator/src/lib/biogrid-state/biogrid-state-class.ts]: Capitalize comment here

File: biogrid/libs/biogrid-simulator/src/lib/biogrid-state/biogrid-state-class.ts
Comment (aosmann7) [reply biogrid/libs/biogrid-simulator/src/lib/biogrid-state/biogrid-state-class.ts:123]: +1

File: biogrid/libs/biogrid-simulator/src/lib/bioenergy-source/bioenergy-source-class.ts
Comment (RolandNaijuka) [reply biogrid/libs/biogrid-simulator/src/lib/bioenergy-source/bioenergy-source-class.ts]: which merge are you referring to? one in master or in this PR

File: biogrid/libs/biogrid-simulator/src/lib/biobattery/biobattery-class.ts
Comment (vitashubin) [inline biogrid/libs/biogrid-simulator/src/lib/biobattery/biobattery-class.ts]: why are we renaming this? getters should be named getVariableName

File: biogrid/libs/biogrid-simulator/src/lib/biobattery/biobattery-class.ts
Comment (vitashubin) [inline biogrid/libs/biogrid-simulator/src/lib/biobattery/biobattery-class.ts:80]: why are we renaming this? getters should be named getVariableName

File: biogrid/libs/biogrid-simulator/src/lib/biobattery/biobattery.spec.ts
Comment (vitashubin) [inline biogrid/libs/biogrid-simulator/src/lib/biobattery/biobattery.spec.ts:30]: instead of having this be a variable string, it looks like we already have a SMALL_BATTERY object. Can we add a "type" or "id" or something similar field into that object and use SMALL_BATTERY.type instead?

File: biogrid/libs/biogrid-simulator/src/lib/biobrain/biobrain-class.ts
Comment (vitashubin) [inline biogrid/libs/biogrid-simulator/src/lib/biobrain/biobrain-class.ts]: nit: can remove extra line

File: biogrid/libs/biogrid-simulator/src/lib/biobrain/biobrain-class.ts
Comment (vitashubin) [inline biogrid/libs/biogrid-simulator/src/lib/biobrain/biobrain-class.ts]: seems like we are never using the full list of grid items. Instead - can this.getGridItems take as a parameter the name which you want, and return only those items? 
e.g. this.getGridItems(GRID_ITEM_NAMES.ENERGY_USER) returns a list of all energy users

**Review by vitashubin — CHANGES_REQUESTED** [review]

File: biogrid/libs/biogrid-simulator/src/lib/biobattery/biobattery-class.ts
Comment (RolandNaijuka) [reply biogrid/libs/biogrid-simulator/src/lib/biobattery/biobattery-class.ts:80]: This is used in the `biobrain` algorithm in method `determineSupplyingPath` to find the energyProvided.
I changed the getter name so that I don't have to check which gridItem it is since all are expected to have this method getEnergyInJoules

File: biogrid/libs/biogrid-simulator/src/lib/biobattery/biobattery-class.ts
Comment (RolandNaijuka) [reply biogrid/libs/biogrid-simulator/src/lib/biobattery/biobattery-class.ts]: This is used in the biobrain algorithm in method determineSupplyingPath to find the energy required by the receiving agent.
I changed the getter name so that I don't have to check which gridItem it is since all are expected to have this getter method MaxCapacity. Also getter methods are different in typescript, you can write `get` then following by the name

File: biogrid/libs/biogrid-simulator/src/lib/biobattery/biobattery.spec.ts
Comment (RolandNaijuka) [reply biogrid/libs/biogrid-simulator/src/lib/biobattery/biobattery.spec.ts:30]: I am writing the name here as is because it is not used in this case. Here it is a placeholder for the `name` requirement in the battery.  The `name` is used for labeling each gridItem on the state graph

File: biogrid/libs/biogrid-simulator/src/lib/biobrain/biobrain-class.ts
Comment (RolandNaijuka) [reply biogrid/libs/biogrid-simulator/src/lib/biobrain/biobrain-class.ts]: This creates an error since we are returning either BioBattery[] or Building[] or BioEnergySource[]. For instance if we want the getGridItems(GRID_ITEM_NAMES.ENERGY_USER), this can technically not only return the energy users but can also return batteries or solar. This cause the compiler to throw an error because we expected to pass energy users not either energy users or batteries..... I will look at a way to implement it using a helper method but there if I change the method, then I have to getGridItems for every type differently.

File: biogrid/libs/biogrid-simulator/src/lib/biogrid/biogrid-class.ts
Comment (RolandNaijuka) [reply biogrid/libs/biogrid-simulator/src/lib/biogrid/biogrid-class.ts:124]: > could this be broken up using helper functions. I.e. if it is an energy user, a function like `takeEnergyUserAction`

will look at how it can be done like this. Thanks for highlighting this out

File: biogrid/libs/biogrid-simulator/src/lib/biogrid/biogrid-class.ts
Comment (RolandNaijuka) [reply biogrid/libs/biogrid-simulator/src/lib/biogrid/biogrid-class.ts:136]: i cannot remove it because a node which has not been changed, will cause an error when I request to start charging it or change it
check lines after the `if..else`

File: biogrid/libs/biogrid-simulator/src/lib/biobrain/biobrain-class.ts
Comment (cwlarson-google) [reply biogrid/libs/biogrid-simulator/src/lib/biobrain/biobrain-class.ts]: You can avoid calling `this.getGridItems()` a bunch of times (abbreviated, but you can extrapolate):

```
const {
  [GRID_ITEM_NAMES.ENERGY_USER]: energyUser,    // JS object deconstruction & renaming
  [GRID_ITEM_NAMES.SMALL_BATTERY]: smallBattery,
} = this.getGridItems();

let buildingSuppliers: SupplyingPath = this.buildingCharging(energyUser, smallBattery);
```

**Review by cwlarson-google — APPROVED** [review]

File: biogrid/libs/biogrid-simulator/src/lib/biobattery/biobattery-class.ts
Comment (vitashubin) [reply biogrid/libs/biogrid-simulator/src/lib/biobattery/biobattery-class.ts:80]: If you change the getter name, you should change the class varialbe name as well.

File: biogrid/libs/biogrid-simulator/src/lib/biobattery/biobattery-class.ts
Comment (vitashubin) [reply biogrid/libs/biogrid-simulator/src/lib/biobattery/biobattery-class.ts]: I understand that this does work, but we want to be consistent in our code. The rest of this code base doesn't use the get <name> method, so we should continue that pattern throughout

File: biogrid/libs/biogrid-simulator/src/lib/biobrain/biobrain-class.ts
Comment (vitashubin) [inline biogrid/libs/biogrid-simulator/src/lib/biobrain/biobrain-class.ts]: nit: capitalize

File: biogrid/libs/biogrid-simulator/src/lib/biobrain/biobrain-class.ts
Comment (vitashubin) [inline biogrid/libs/biogrid-simulator/src/lib/biobrain/biobrain-class.ts]: nit: "energy" written twice

File: biogrid/libs/biogrid-simulator/src/lib/biobrain/biobrain-class.ts
Comment (vitashubin) [inline biogrid/libs/biogrid-simulator/src/lib/biobrain/biobrain-class.ts]: nit: rename to "chargeBuildings" to be consistent with "chargeSmallBatteries"

File: biogrid/libs/biogrid-simulator/src/lib/biobrain/biobrain-class.ts
Comment (vitashubin) [inline biogrid/libs/biogrid-simulator/src/lib/biobrain/biobrain-class.ts:123]: do you need this line of code? isn't solarPanels already an array?

File: biogrid/libs/biogrid-simulator/src/lib/biobrain/biobrain-class.ts
Comment (vitashubin) [inline biogrid/libs/biogrid-simulator/src/lib/biobrain/biobrain-class.ts:201]: what do you mean `the grid item without it`?

File: biogrid/libs/biogrid-simulator/src/lib/biobrain/biobrain-class.ts
Comment (vitashubin) [inline biogrid/libs/biogrid-simulator/src/lib/biobrain/biobrain-class.ts:202]: nit: s/enrgy/energy

File: biogrid/libs/biogrid-simulator/src/lib/biobrain/biobrain-class.ts
Comment (vitashubin) [inline biogrid/libs/biogrid-simulator/src/lib/biobrain/biobrain-class.ts:208]: can you clarify what the key and value pair is?

File: biogrid/libs/biogrid-simulator/src/lib/biobrain/biobrain-class.ts
Comment (vitashubin) [inline biogrid/libs/biogrid-simulator/src/lib/biobrain/biobrain-class.ts:219]: not sure I understand this comment - what do you mean so that it can not be requested energy when it doesn't have it?

File: biogrid/libs/biogrid-simulator/src/lib/biobrain/biobrain-class.ts
Comment (vitashubin) [inline biogrid/libs/biogrid-simulator/src/lib/biobrain/biobrain-class.ts]: can we file a ticket to make it not all or nothing and tag this as a TODO?

File: biogrid/libs/biogrid-simulator/src/lib/biobrain/biobrain-class.ts
Comment (vitashubin) [inline biogrid/libs/biogrid-simulator/src/lib/biobrain/biobrain-class.ts]: instead of calling the variable directly, we should be calling a getter method (should be `receivingAgent.getMaxCapacity()`

Comment (vitashubin) [reply biogrid/libs/biogrid-simulator/src/lib/biobrain/biobrain-class.ts]: I understand that this works as is, but prefer using getter methods and staying consistent

File: biogrid/libs/biogrid-simulator/src/lib/biobrain/biobrain.spec.ts
Comment (vitashubin) [inline biogrid/libs/biogrid-simulator/src/lib/biobrain/biobrain.spec.ts]: nit: remove extra space

File: biogrid/libs/biogrid-simulator/src/lib/biobrain/biobrain.spec.ts
Comment (vitashubin) [inline biogrid/libs/biogrid-simulator/src/lib/biobrain/biobrain.spec.ts]: nit: rephrase and remove "it should"

Comment (vitashubin) [reply biogrid/libs/biogrid-simulator/src/lib/biobrain/biobrain.spec.ts]: I'll talk about this in stand up today as well

File: biogrid/libs/biogrid-simulator/src/lib/bioenergy-source/solar-panel.ts
Comment (vitashubin) [inline biogrid/libs/biogrid-simulator/src/lib/bioenergy-source/solar-panel.ts]: why not just check if this.getEnergyInJoules === 0?

**Review by vitashubin — CHANGES_REQUESTED** [review]

File: biogrid/libs/biogrid-simulator/src/lib/biobattery/biobattery-class.ts
Comment (Lev-Stambler) [reply biogrid/libs/biogrid-simulator/src/lib/biobattery/biobattery-class.ts]: I agree with Vita on the getters. Otherwise, all looks good pending changes

**Review by Lev-Stambler — APPROVED** [review]
Approved pending changes

File: biogrid/libs/biogrid-simulator/src/lib/biobattery/biobattery-class.ts
Comment (vitashubin) [inline biogrid/libs/biogrid-simulator/src/lib/biobattery/biobattery-class.ts:14]: please rename to batteryType to be more specific 
Additionally,  since there are limited options for possible types it could be, if instead of a string we can have this represented as an enum value, that would be even better :)

File: biogrid/libs/biogrid-simulator/src/lib/biobrain/biobrain-class.ts
Comment (vitashubin) [inline biogrid/libs/biogrid-simulator/src/lib/biobrain/biobrain-class.ts:38]: looks like you are passing in all the gridItems anyway, why not pass the full gridItems list to tihs.chargeBuildings, and then let chargeBuildings differentiate between item types?

File: biogrid/libs/biogrid-simulator/src/lib/bioenergy-source/solar-panel.ts
Comment (vitashubin) [inline biogrid/libs/biogrid-simulator/src/lib/bioenergy-source/solar-panel.ts:37]: is this supposed to be a unique name? looks like it defaults to GRID_ITEM_NAMES.SOLAR_PANEL, what happens if multiple panels have the same name? is that okay?

File: biogrid/libs/biogrid-simulator/src/lib/biogrid-state/biogrid-state-class.ts
Comment (vitashubin) [inline biogrid/libs/biogrid-simulator/src/lib/biogrid-state/biogrid-state-class.ts:60]: update comment or remove

File: biogrid/libs/biogrid-simulator/src/lib/config/bio-constants.ts
Comment (vitashubin) [inline biogrid/libs/biogrid-simulator/src/lib/config/bio-constants.ts:17]: what happens if they go below this min?

**Review by vitashubin — COMMENTED** [review]
Please address remaining comments in a follow-up PR :) Thanks!

File: biogrid/libs/biogrid-simulator/src/lib/biobattery/biobattery-class.ts
Comment (RolandNaijuka) [reply biogrid/libs/biogrid-simulator/src/lib/biobattery/biobattery-class.ts:14]: This name is used for the vertices on the graph. It is unique to every item. For instance if we have 3 small batteries on the grid, all of them have different name variable. I will change it to `gridItemName` as it makes it easy to understand. I hope this answers the question

File: biogrid/libs/biogrid-simulator/src/lib/biobrain/biobrain-class.ts
Comment (RolandNaijuka) [reply biogrid/libs/biogrid-simulator/src/lib/biobrain/biobrain-class.ts:38]: I will address that, thank you

File: biogrid/libs/biogrid-simulator/src/lib/biogrid-state/biogrid-state-class.ts
Comment (RolandNaijuka) [reply biogrid/libs/biogrid-simulator/src/lib/biogrid-state/biogrid-state-class.ts:60]: I will update the comments, thanks for highlighting this out

File: biogrid/libs/biogrid-simulator/src/lib/bioenergy-source/solar-panel.ts
Comment (RolandNaijuka) [reply biogrid/libs/biogrid-simulator/src/lib/bioenergy-source/solar-panel.ts:37]: It is unique, I implemented it with a default for cases where there is just one solar panel. One doesn't have to pass the name, but I will make it compulsory since it will be unique at the end of the day.

File: biogrid/libs/biogrid-simulator/src/lib/config/bio-constants.ts
Comment (RolandNaijuka) [reply biogrid/libs/biogrid-simulator/src/lib/config/bio-constants.ts:17]: it starts to wear down and degrades the panels

**Pull request merged**
